File: src/main/java/com/lessmarkup/framework/data/EditableCollectionCache.java
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */
package com.lessmarkup.framework.data;

import com.lessmarkup.interfaces.cache.AbstractCacheHandler;
import com.lessmarkup.interfaces.data.DataObject;
import com.lessmarkup.interfaces.recordmodel.RecordModel;
import com.lessmarkup.interfaces.structure.Tuple;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Component;

import java.lang.reflect.Method;
import java.lang.reflect.Modifier;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.OptionalLong;

@Component
@Scope("prototype")
public class EditableCollectionCache extends AbstractCacheHandler {

    private final Map<Tuple<Class<? extends RecordModel>, Class<? extends DataObject>>, List<RecordToDataPropertyMapper>> propertySets = new HashMap<>();
    
    public EditableCollectionCache() {
        super(null);
    }
    
    @Override
    public void initialize(OptionalLong objectId) {
        if (objectId.isPresent()) {
            throw new IllegalArgumentException();
        }
    }
    
    private static Map<String, Method> getTypeMethods(Class<?> type) {
        Map<String, Method> methods = new HashMap<>();
        for (Method method : type.getMethods()) {
            if ((method.getModifiers() & Modifier.STATIC) != 0) {
                continue;
            }
            methods.put(method.getName(), method);
        }
        return methods;
    }
    
    synchronized List<RecordToDataPropertyMapper> getProperties(Class<? extends RecordModel> modelType, Class<? extends DataObject> dataType) {
        
        Tuple<Class<? extends RecordModel>, Class<? extends DataObject>> key = new Tuple(modelType, dataType);
        
        List<RecordToDataPropertyMapper> ret = propertySets.get(key);
        if (ret != null) {
            return ret;
        }
        
        ret = new ArrayList<>();
        
        Map<String, Method> modelMethods = getTypeMethods(modelType);
        Map<String, Method> dataMethods = getTypeMethods(dataType);
        
        for (Entry<String, Method> entry : modelMethods.entrySet()) {
            if (!entry.getKey().startsWith("set")) {
                continue;
            }
            
            String setterName = entry.getKey();
            Method modelSetter = entry.getValue();
            
            if (modelSetter.getParameterCount() != 1) {
                continue;
            }
            
            Class<?> fieldType = modelSetter.getParameterTypes()[0];
            
            String getterName;
            
            if (fieldType.equals(boolean.class)) {
                getterName = "is" + setterName.substring(3);
            } else {
                getterName = "get" + setterName.substring(3);
            }
            
            Method modelGetter = modelMethods.get(getterName);
            if (modelGetter == null || modelGetter.getParameterCount() != 0 || !modelGetter.getReturnType().equals(fieldType)) {
                continue;
            }
            
            Method dataSetter = dataMethods.get(setterName);
            if (dataSetter == null || dataSetter.getParameterCount() != 1 || !dataSetter.getParameterTypes()[0].equals(fieldType)) {
                continue;
            }
            
            Method dataGetter = dataMethods.get(getterName);
            if (dataGetter == null || dataGetter.getParameterCount() != 0 || !dataGetter.getReturnType().equals(fieldType)) {
                continue;
            }
            
            ret.add(new RecordToDataPropertyMapper(dataGetter, dataSetter, modelGetter, modelSetter, fieldType));
        }
        
        propertySets.put(key, ret);
        
        return ret;
    }
}

File: src/main/scala/com/lessmarkup/framework/data/EditableCollectionCache.scala
/*
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/.
 */

package com.lessmarkup.framework.data

import java.lang.reflect.{Method, Modifier}

import com.lessmarkup.interfaces.cache.AbstractCacheHandler
import com.lessmarkup.interfaces.data.DataObject
import com.lessmarkup.interfaces.recordmodel.RecordModel

import scala.collection.mutable

object EditableCollectionCache {
  private def getTypeMethods(modelType: Class[_]): Map[String, Method] = {
    modelType.getMethods.filter(m => Modifier.isStatic(m.getModifiers)).map(m => (m.getName, m)).toMap
  }
}

class EditableCollectionCache extends AbstractCacheHandler {
  private final val propertySets: mutable.HashMap[(Class[_ <: RecordModel[_]], Class[_ <: DataObject]), List[RecordToDataPropertyMapper]] = mutable.HashMap()

  def getProperties(modelType: Class[_ <: RecordModel[_]], dataType: Class[_ <: DataObject]): List[RecordToDataPropertyMapper] = {
    val key = (modelType, dataType)
    val ret: Option[List[RecordToDataPropertyMapper]] = propertySets.get(key)
    if (ret.isDefined) {
      return ret.get
    }

    val modelMethods: Map[String, Method] = EditableCollectionCache.getTypeMethods(modelType)
    val dataMethods: Map[String, Method] = EditableCollectionCache.getTypeMethods(dataType)

    val properties = modelMethods
      .filter(m => m._1.startsWith("set") && m._2.getParameterCount == 1)
      .map {
        case (setterName, modelSetter) =>
          val fieldType: Class[_] = modelSetter.getParameterTypes()(0)
          var getterName: String = null
          if (fieldType == classOf[Boolean]) {
            getterName = "is" + setterName.substring(3)
          }
          else {
            getterName = "get" + setterName.substring(3)
          }

          val modelGetter: Option[Method] = modelMethods.get(getterName)
          val dataSetter: Option[Method] = dataMethods.get(setterName)
          val dataGetter: Option[Method] = dataMethods.get(getterName)

          if (modelGetter.isEmpty || modelGetter.get.getParameterCount != 0 || !(modelGetter.get.getReturnType == fieldType)
            || dataSetter.isEmpty || dataSetter.get.getParameterCount != 1 || !(dataSetter.get.getParameterTypes()(0) == fieldType)
            || dataGetter.isEmpty || dataGetter.get.getParameterCount != 0 || !(dataGetter.get.getReturnType == fieldType)) {
            None
          } else {
            Option(new RecordToDataPropertyMapper(dataGetter.get, dataSetter.get, modelGetter.get, modelSetter, fieldType))
          }
    }
      .filter(_.isDefined)
      .map(_.get)
      .toList

    propertySets.put(key, properties)
    properties
  }
}
